sig-scalability/blogs/k8s-services-scalability-issues.md

```diff
@@ -0,0 +1,117 @@
+# Known Scalability Issues with Kubernetes Services
+
+_by Shyam JVS, Google Inc (with inputs from Brian Grant & Wojciech Tyczynski)_
+
+**June 2018**
+
+This document serves as a catalog of issues we've known/discovered with kubernetes services as of June 2018, focusing on their scalability/performance. The purpose of the document is to make the information common knowledge for the community, so we can work together towards improving it. Listing them below in no particular order.
+
+
+## Performance of iptables when large number of service ports exist
+
+#### Issue
+
+Iptables can be slow in the presence of a large number of services. As the number of service ports increases, the KUBE-SERVICES chain gets longer and since it’s also evaluated very frequently that can bring down the performance. Also, it seems that manipulating a chain with lots of rules (e.g. KUBE-SERVICES) takes noticeably longer than others.
+
+#### Relevant issues/links
+
+- [#48938](https://github.com/kubernetes/kubernetes/issues/48938)
+
+#### Possible Solution(s)
+
+- Moving from iptables to IPVS may help here if that works stably. We have the IPVS alternative [implemented](https://github.com/kubernetes/kubernetes/pull/46580), but it still hasn’t gone to GA. Also we need to measure/confirm performance benefits from it.
```

Comment:
For a long time the official kernel upstream answer was that nftables was going to solve all iptables-related scalability problems (both with packet processing and with rule changes), and there's now an out-of-tree nftables kube-proxy backend too. (https://github.com/kubernetes/kubernetes/issues/62720).

There's also an alternative plan to fix the packet processing speed problems by rewriting iptables to use eBPF inside the kernel (https://cilium.io/blog/2018/04/17/why-is-the-kernel-community-replacing-iptables/). (That doesn't necessarily address the reload-speed problem though, and there's a limit to how much that can be fixed without making incompatible changes to the userspace iptables "API".)

Reply:
I've incorporated those points - PTAL. Also thanks for those really useful leads!